File: scripts/workflow_guard.py
#!/usr/bin/env python3
"""Guard Coding Plugins workflow transitions from document-chain state."""

from __future__ import annotations

import argparse
import json
from pathlib import Path
from typing import Any

import workflow_state


VALID_TARGETS = {"plan", "execute"}


def check(root: Path | str, *, feature: str, doc_id: str, target: str) -> dict[str, Any]:
    if target not in VALID_TARGETS:
        raise ValueError(f"invalid workflow guard target: {target}")

    state = workflow_state.inspect_document_chain(root, feature=feature, doc_id=doc_id)
    allowed_states = {
        "plan": {"ready-for-plan"},
        "execute": {"ready-for-execution"},
    }[target]
    passed = state["state"] in allowed_states and not state["stale"]

    failures: list[str] = []
    if not passed:
        if state["missing_artifacts"]:
            failures.append(f"missing artifacts: {', '.join(state['missing_artifacts'])}")
        if state["state"] == "plan-unlocked":
            failures.append("IPD source_hash is missing")
        if state["stale"]:
            failures.append("IPD source_hash is stale")
        if state["state"] not in allowed_states:
            failures.append(f"state '{state['state']}' cannot enter target '{target}'")

    return {
        "pass": passed,
        "target": target,
        "feature": feature,
        "doc_id": doc_id,
        "state": state["state"],
        "next_skill": state["next_skill"],
        "reason": state["reason"],
        "missing_artifacts": state["missing_artifacts"],
        "stale": state["stale"],
        "failures": failures,
    }


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="Check Coding Plugins workflow guard.")
    parser.add_argument("command", choices=("check",))
    parser.add_argument("--root", default=".")
    parser.add_argument("--feature", required=True)
    parser.add_argument("--doc-id", required=True)
    parser.add_argument("--target", required=True, choices=sorted(VALID_TARGETS))
    parser.add_argument("--json", action="store_true")
    return parser


def main(argv: list[str] | None = None) -> int:
    args = build_parser().parse_args(argv)
    result = check(Path(args.root), feature=args.feature, doc_id=args.doc_id, target=args.target)
    if args.json:
        print(json.dumps(result, ensure_ascii=False, indent=2))
    else:
        print(f"pass: {str(result['pass']).lower()}")
        print(f"target: {result['target']}")
        print(f"state: {result['state']}")
        print(f"next_skill: {result['next_skill']}")
        print(f"reason: {result['reason']}")
        if result["failures"]:
            print("failures:")
            for failure in result["failures"]:
                print(f"- {failure}")
    return 0 if result["pass"] else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: src/cli/workflow/workflow-guard.ts
#!/usr/bin/env node
import { resolve } from "node:path";

import { checkWorkflowGuard, VALID_TARGETS } from "../../lib/workflow/workflow-guard.ts";

interface Options {
  command?: "check";
  root: string;
  feature?: string;
  docId?: string;
  target?: string;
  json: boolean;
}

function parseArgs(argv: string[]): Options {
  const [command, ...rest] = argv;
  if (command !== "check") {
    throw new Error("command must be check.");
  }
  const options: Options = { command, root: ".", json: false };

  for (let index = 0; index < rest.length; index += 1) {
    const arg = rest[index];
    if (arg === "--root") {
      options.root = requireValue(rest, index, arg);
      index += 1;
    } else if (arg === "--feature") {
      options.feature = requireValue(rest, index, arg);
      index += 1;
    } else if (arg === "--doc-id") {
      options.docId = requireValue(rest, index, arg);
      index += 1;
    } else if (arg === "--target") {
      options.target = requireValue(rest, index, arg);
      if (!VALID_TARGETS.has(options.target)) {
        throw new Error(`invalid workflow guard target: ${options.target}`);
      }
      index += 1;
    } else if (arg === "--json") {
      options.json = true;
    } else {
      throw new Error(`Unknown argument: ${arg}`);
    }
  }

  if (!options.feature) {
    throw new Error("--feature is required.");
  }
  if (!options.docId) {
    throw new Error("--doc-id is required.");
  }
  if (!options.target) {
    throw new Error("--target is required.");
  }
  options.root = resolve(options.root);
  return options;
}

function requireValue(argv: string[], index: number, arg: string): string {
  const value = argv[index + 1];
  if (!value) {
    throw new Error(`${arg} requires a value.`);
  }
  return value;
}

try {
  const options = parseArgs(process.argv.slice(2));
  const result = checkWorkflowGuard(options.root, {
    feature: options.feature ?? "",
    docId: options.docId ?? "",
    target: options.target ?? "",
  });

  if (options.json) {
    console.log(JSON.stringify(result, null, 2));
  } else {
    console.log(`pass: ${String(result.pass).toLowerCase()}`);
    console.log(`target: ${result.target}`);
    console.log(`state: ${result.state}`);
    console.log(`next_skill: ${result.next_skill}`);
    console.log(`reason: ${result.reason}`);
    if (result.failures.length > 0) {
      console.log("failures:");
      for (const failure of result.failures) {
        console.log(`- ${failure}`);
      }
    }
  }
  process.exit(result.pass ? 0 : 1);
} catch (error) {
  const message = error instanceof Error ? error.message : String(error);
  console.error(message);
  process.exit(1);
}
